src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs

```diff
@@ -95,13 +95,35 @@ internal partial class CSharpIntroduceVariableService
             var rewrittenBody = Rewrite(
                 document, expression, newLocalName, document, oldBody, allOccurrences, cancellationToken);
 
+            var doesDelegateMethodReturnVoid =
+                document.SemanticModel.GetTypeInfo(oldLambda, cancellationToken).ConvertedType is INamedTypeSymbol delegateType &&
+                delegateType.DelegateInvokeMethod != null &&
+                delegateType.DelegateInvokeMethod.ReturnsVoid;
 
-            var newBody =
-                document.SemanticModel.GetTypeInfo(oldLambda, cancellationToken).ConvertedType is INamedTypeSymbol delegateType
-                && delegateType.DelegateInvokeMethod != null
-                && delegateType.DelegateInvokeMethod.ReturnsVoid
-                    ? SyntaxFactory.Block(declarationStatement)
-                    : SyntaxFactory.Block(declarationStatement, SyntaxFactory.ReturnStatement(rewrittenBody));
+            BlockSyntax newBody;
+            if (doesDelegateMethodReturnVoid)
+            {
+                // We don't need to include the rewritten body if it consists of just the identifier name.
+                // However, in all other cases, the rewritten body should be included.
+                var isRewrittenBodyTrivial = rewrittenBody.WalkDownParentheses().IsKind(SyntaxKind.IdentifierName,
+                    out IdentifierNameSyntax identifier) &&
+                    identifier.Identifier.ValueText == newLocalName.Identifier.ValueText;
+
+                if (isRewrittenBodyTrivial)
+                {
+                    newBody = SyntaxFactory.Block(declarationStatement);
```

Comment:
this portion here (from 106-114) seems like some sort of hack to get some old behavior.  could you clarify overall what hte issue is and what sorts of before/after cases we want to be supporting and why?

Reply:
Previously, if the lambda had a void return type, the introduce local refactoring deleted the original code in the lambda statement and only kept the new declaration statement.
e.g. In this case:
```C#
Task.Run(() => File.Copy("src", [|Path.Combine("dir", "file")|]));
```
The refactoring incorrectly turned it into:
```C#
Task.Run(() =>
{
    string destFileName = Path.Combine("dir", "file");
});
```


I believe the original code was written to aid in cases such as:
```C#
Action<int> goo = x => [|x.ToString()|];
```
which, without the original check, would have ended up like:
```C#
Action<int> goo = x =>
{
    string v = x.ToString();
    v;
};
```

The updated code essentially checks to see if the code being replaced with the new variable is trivial or not. (i.e., is it just the identifier? -> if so, we don't need to include it.)

Reply:
> The updated code essentially checks to see if the code being replaced with the new variable is trivial or not. (i.e., is it just the identifier? -> if so, we don't need to include it.)

Then what *do* we do for:

```Action<int> goo = x => [|x.ToString()|];```?

What do we generate in that case?

Reply:
The current functionality generates the following for the case, which I believe is correct:
```C#
        Action<int> goo = x =>
        {
            string v = x.ToString();
        };
```

Reply:
ok, that makes sense (and matches what we do with normal methods).  

i'll have to noodle on this a bit.  something doesn't feel really clear ot me here.  

If anything, have you considered having this code just pass a value into the Rewrite function saying it's a void/non-void context and if the expression is being returned or or not?  It might just be simpler to have the rewrite function try to encode this logic instead of effectively "undoing" it here at this point.

Reply:
I definitely see where you're coming from regarding passing a value into the Rewrite function. I personally think it may make the code a bit messier since this is a C#-specific issue and the Rewrite function is language-agnostic. 

The language-specific method Rewrite calls into doesn't take in the appropriate parameters, so if we decided to add parameters, they would have to be added to the C# and VB implementations. I'm not sure if this would be considered too much clutter, since VB wouldn't be using the added parameters. Or we could add a new abstract method and the VB implementation could return nothing. Do you have a perspective on what approach you think might be best?

Reply:
fair point about the shared Vb/C# logic.

what i really don't like is having some function return data to us that we then have to break apart.  especially with realy janky code like:

```c#
                var isRewrittenBodyTrivial = rewrittenBody.WalkDownParentheses().IsKind(SyntaxKind.IdentifierName,
                    out IdentifierNameSyntax identifier) &&
                    identifier.Identifier.ValueText == newLocalName.Identifier.ValueText;
```

if we need to know if the new body is trivial, why can't that be something that the `Rewrite` function returns out to us as a bool since it must have known this when creating the new body in the first place?

--

Basically, we shouldn't have code that has to re-infer information about what's going on from code deeper in the exact same component.  It indicates a band-aid on a problem after the fact instead of an appropriate amount of work at the source to figure out what's going on.  

So if possible, have `Rewrite` figure things out and return enough real information for the higher up caller to do the right hting.  For VB we can likely just ignore the information (or can we?) if it's not doing special stuff with lambdas.  But for c# we can use that extra information to drive a more suitable experience at this layer.